File: src/opaque/pose/RefNode.py

class RefNode:

	def __init__(self, nid, jid, x, z, p, probe):

		self.refX = x
		self.refZ = z
		self.refP = p

		self.gndX = x
		self.gndZ = z
		self.gndP = p

		self.nodeID = nid
		self.jointID = jid

		self.probe = probe

		self.numJoints = self.probe.getNumJoints()

		self.nom_ang = [0.0 for i in range(self.numJoints)]
		for i in range(self.numJoints):
			self.nom_ang[i] = probe.getServo(i)

		self.maxError = 0.
		self.active_time = 0
		self.errSum = 0.
		self.currError = 0.
		self.maxErrorReached = False
		self.newRoot = False


	def getRefPose(self):
		return self.refX, self.refZ, self.refP

	def getNodeID(self):
		return self.nodeID

	def getJointID(self):
		return self.jointID

	def setGroundTruthPose(self, x, z, p):
		self.gndX = x
		self.gndZ = z
		self.gndP = p

	def getGroundTruthPose(self):
		return self.gndX, self.gndZ, self.gndP

	def getNumEdges(self):
		return len(self.edges)

	def computeStabilityError(self):
	
		# saddle points: 4,8,12,16,20,24,28,32,36
		self.interval = self.jointID % 4

		# find the saddle point joints to which to check stability
		if self.interval == 0:
			self.low_index = self.jointID
			self.high_index = self.jointID + 4

		else:

			self.low_index = self.jointID - self.interval
			self.high_index = self.low_index + 4
		
		# set the max and min limits if they go over
		if self.low_index < 0:
			self.low_index = 0

		if self.high_index >= self.numJoints:
			self.high_index = self.numJoints - 1

		error = 0.0
		for i in range(self.low_index,self.high_index + 1):
			error += abs(self.nom_ang[i] - self.probe.getServo(i))

		if error > self.maxError:
			self.maxError = error

		self.currError = error
		self.errSum += error

		if self.currError > 0.1:
			self.maxErrorReached = True

		return error

	def getStabilityError(self):
		return self.currError

	def getMaxStabilityError(self):
		return self.maxError

	def getAvgStabilityError(self):
		return self.errSum/self.active_time

	def isMaxErrorReached(self):
		return self.maxErrorReached

	def updateTime(self):
		self.active_time += 1

	def setNewRoot(self, isRoot):
		self.newRoot = isRoot

	def isNewRoot(self):
		return self.newRoot

File: src/RefNode.cpp
#include "RefNode.h"
#include <cmath>

RefNode::RefNode(int nid, int jid, double x, double z, double p, int numJoints, double *joints) {

	refX = x;
	refZ = z;
	refP = p;

	gndX = x;
	gndZ = z;
	gndP = p;

	nodeID = nid;
	jointID = jid;
	this->numJoints = numJoints;

	for (int i = 0 ; i < numJoints ; i++ ) {
		nom_ang[i] = joints[i];
	}

	maxError = 0.0;
	active_time = 0;
	errSum = 0.0;
	currError = 0.0;
	maxErrorReached = 0;
	newRoot = 0;

}

RefNode::~RefNode() {

}

double RefNode::getRefPoseX() {
	return refX;
}

double RefNode::getRefPoseZ() {
	return refZ;
}

double RefNode::getRefPoseP() {
	return refP;
}

int RefNode::getNodeID() {
	return nodeID;
}

int RefNode::getJointID() {
	return jointID;
}

void RefNode::setGroundTruthPose(double x, double z, double p) {
	gndX = x;
	gndZ = z;
	gndP = p;
}

double RefNode::getGroundTruthPoseX() {
	return gndX;
}

double RefNode::getGroundTruthPoseZ() {
	return gndZ;
}

double RefNode::getGroundTruthPoseP() {
	return gndP;
}

double RefNode::computeStabilityError(double *joints) {

	// saddle points: 4,8,12,16,20,24,28,32,36
	int interval = jointID % 4;
	int low_index, high_index;

	// find the saddle point joints to which to check stability
	if (interval == 0) {
		low_index = jointID;
		high_index = jointID + 4;
	}
	else {
		low_index = jointID - interval;
		high_index = low_index + 4;
	}
	
	// set the max and min limits if they go over
	if (low_index < 0) 
		low_index = 0;

	if (high_index >= numJoints)
		high_index = numJoints - 1;

	double error = 0.0;
	for (int i = low_index; i <= high_index ; i++ ) 
		error += fabs(nom_ang[i] - joints[i]);

	if (error > maxError)
		maxError = error;

	currError = error;
	errSum += error;

	if (currError > 0.1)
		maxErrorReached = 1; 

	return error;

}

double RefNode::getStabilityError() {
	return currError;
}

double RefNode::getMaxStabilityError() {
	return maxError;
}

double RefNode::getAvgStabilityError() {
	return errSum/active_time;
}

int RefNode::isMaxErrorReached() {
	return maxErrorReached;
}

void RefNode::updateTime() {
	active_time += 1;
}

void RefNode::setNewRoot(int isRoot) {
	newRoot = isRoot;
}

int RefNode::isNewRoot() {
	return newRoot;
}



